[package] <cyclonedds-cxx>/<0.10.4>: building error when enable cyclonedds option with_shm=True

### Description

The issue occurs when using cyclonedds-cxx in a conanfile, and enabling the cyclonedds option with_shm=True. The compilation fails.
conanfile.py
```
from conan import ConanFile
from conan.tools.cmake import cmake_layout


class ExampleRecipe(ConanFile):
    settings = "os", "compiler", "build_type", "arch"
    generators = "CMakeDeps", "CMakeToolchain"

    def configure(self):
        self.options["cyclonedds"].with_shm = True

    def requirements(self):
        self.requires("cyclonedds-cxx/0.10.4")

    def layout(self):
        cmake_layout(self)
```

### Package and Environment Details

* Package Name/Version: **cyclonedds-cxx/0.10.4**
* Operating System+version: **Linux Ubuntu 20.04**
* Compiler+version: **GCC 9**
* Conan version: **conan 1.61.0**


### Conan profile

[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux
os_build=Linux
[options]
[build_requires]
[env]
[conf]
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True

### Steps to reproduce

mkdir build && cd build
conan install .. --build=missing

### Logs

<details><summary>Click to expand log</summary>

```
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux
os_build=Linux
[options]
[build_requires]
[env]
[conf]
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True

cyclonedds-cxx/0.10.4: WARN: Package binary is corrupted, removing: 223cf2061ff02a5cb199bdb3c175c5cff1c27086
Version ranges solved
    Version range '>=3.16 <4' required by 'cyclonedds-cxx/0.10.4' resolved to 'cmake/3.30.5' in local cache

conanfile.py: Installing package
Requirements
    acl/2.3.1 from 'mcd' - Cache
    cpptoml/0.1.1 from 'conancenter' - Cache
    cyclonedds/0.10.4 from 'conancenter' - Cache
    cyclonedds-cxx/0.10.4 from 'conancenter' - Cache
    iceoryx/2.0.5 from 'conancenter' - Cache
    libattr/2.5.1 from 'mcd' - Cache
Packages
    acl/2.3.1:2663adcc87e209148db732c94c276dbfa8a77882 - Cache
    cpptoml/0.1.1:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    cyclonedds/0.10.4:07f3d1230b26cb7f4b51f259cc8b3fcc8a47e074 - Cache
    cyclonedds-cxx/0.10.4:223cf2061ff02a5cb199bdb3c175c5cff1c27086 - Build
    iceoryx/2.0.5:f5eb72abcfbcb55e442970061a2b94a80403d80f - Cache
    libattr/2.5.1:6af9cc7cb931c5ad942174fd7838eb655717c709 - Cache
Build requirements
    cmake/3.30.5 from 'mcd' - Cache
Build requirements packages
    cmake/3.30.5:4db1be536558d833e52e862fd84d64d75c2b3656 - Cache

Installing (downloading, building) binaries...
cmake/3.30.5: Already installed!
cmake/3.30.5: Appending PATH environment variable: /data/conan/data/cmake/3.30.5/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/bin
cpptoml/0.1.1: Already installed!
libattr/2.5.1: Already installed!
acl/2.3.1: Already installed!
iceoryx/2.0.5: Already installed!
iceoryx/2.0.5: Appending PATH environment variable: /data/conan/data/iceoryx/2.0.5/_/_/package/f5eb72abcfbcb55e442970061a2b94a80403d80f/bin
cyclonedds/0.10.4: Already installed!
cyclonedds-cxx/0.10.4: Applying build-requirement: cmake/3.30.5
cyclonedds-cxx/0.10.4: WARN: Build folder is dirty, removing it: /data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086
cyclonedds-cxx/0.10.4: Copying sources to build folder
cyclonedds-cxx/0.10.4: Building your package in /data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086
cyclonedds-cxx/0.10.4: Generator txt created conanbuildinfo.txt
cyclonedds-cxx/0.10.4: Calling generate()
cyclonedds-cxx/0.10.4: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
cyclonedds-cxx/0.10.4: Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
cyclonedds-cxx/0.10.4: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
cyclonedds-cxx/0.10.4: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
cyclonedds-cxx/0.10.4: Aggregating env generators
cyclonedds-cxx/0.10.4: Calling build()
cyclonedds-cxx/0.10.4: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/data/conan/data/cyclonedds-cxx/0.10.4/_/_/package/223cf2061ff02a5cb199bdb3c175c5cff1c27086" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086/src"
-- Using Conan toolchain: /data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Component target declared 'CycloneDDS::ddsc'
-- Conan: Component target declared 'CycloneDDS::idl'
-- Conan: Target declared 'CycloneDDS::CycloneDDS'
-- Conan: Component target declared 'iceoryx_hoofs::iceoryx_platform'
-- Conan: Component target declared 'iceoryx_hoofs::iceoryx_hoofs'
-- Conan: Component target declared 'iceoryx_posh::iceoryx_posh'
-- Conan: Component target declared 'iceoryx_posh::iceoryx_posh_roudi'
-- Conan: Component target declared 'iceoryx_posh::iceoryx_posh_gateway'
-- Conan: Component target declared 'iceoryx_posh::iceoryx_posh_config'
-- Conan: Component target declared 'iceoryx_binding_c::iceoryx_binding_c'
-- Conan: Target declared 'iceoryx::iceoryx'
-- Conan: Target declared 'acl::acl'
-- Conan: Target declared 'libattr::libattr'
-- Conan: Target declared 'cpptoml'
-- Conan: Including build module from '/data/conan/data/cyclonedds/0.10.4/_/_/package/07f3d1230b26cb7f4b51f259cc8b3fcc8a47e074/lib/cmake/CycloneDDS/CycloneDDS_idlc.cmake'
-- Conan: Including build module from '/data/conan/data/cyclonedds/0.10.4/_/_/package/07f3d1230b26cb7f4b51f259cc8b3fcc8a47e074/lib/cmake/CycloneDDS/idlc/Generate.cmake'
CMake Error at CMakeLists.txt:227 (find_package):
  By not providing "Findiceoryx_binding_c.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "iceoryx_binding_c", but CMake did not find one.

  Could not find a package configuration file provided by "iceoryx_binding_c"
  with any of the following names:

    iceoryx_binding_cConfig.cmake
    iceoryx_binding_c-config.cmake

  Add the installation prefix of "iceoryx_binding_c" to CMAKE_PREFIX_PATH or
  set "iceoryx_binding_c_DIR" to a directory containing one of the above
  files.  If "iceoryx_binding_c" provides a separate development package or
  SDK, be sure it has been installed.


-- Configuring incomplete, errors occurred!
cyclonedds-cxx/0.10.4:
cyclonedds-cxx/0.10.4: ERROR: Package '223cf2061ff02a5cb199bdb3c175c5cff1c27086' build failed
cyclonedds-cxx/0.10.4: WARN: Build folder /data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086/build/Release
ERROR: cyclonedds-cxx/0.10.4: Error in build() method, line 119
        cmake.configure()
        ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/data/conan/data/cyclonedds-cxx/0.10.4/_/_/package/223cf2061ff02a5cb199bdb3c175c5cff1c27086" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/data/conan/data/cyclonedds-cxx/0.10.4/_/_/build/223cf2061ff02a5cb199bdb3c175c5cff1c27086/src"
```

</details>

--- a/recipes/cyclonedds-cxx/all/conanfile.py
+++ b/recipes/cyclonedds-cxx/all/conanfile.py
@@ -21,10 +21,12 @@ class CycloneDDSCXXConan(ConanFile):
     options = {
         "shared": [True, False],
         "fPIC": [True, False],
+        "with_shm": [True, False],
     }
     default_options = {
         "shared": False,
         "fPIC": True,
+        "with_shm": False,
     }
 
     @property
@@ -69,6 +71,9 @@ class CycloneDDSCXXConan(ConanFile):
         #      <dds/topic/detail/Topic.hpp>:34
         self.requires("cyclonedds/{}".format(self.version), transitive_headers=True)
 
+        if self.options.with_shm:
+            self.requires("iceoryx/2.0.5")
+
     def validate(self):
         if self.settings.compiler.get_safe("cppstd"):
             check_min_cppstd(self, self._min_cppstd)
@@ -77,6 +82,10 @@ class CycloneDDSCXXConan(ConanFile):
             raise ConanInvalidConfiguration(
                 f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
             )
+
+        if self.options.with_shm != self.dependencies['cyclonedds'].options.with_shm:
+            raise ConanInvalidConfiguration(
+                "cyclonedds-cxx and cyclonedds must be built with the same 'with_shm' option")
 
     def build_requirements(self):
         self.tool_requires("cmake/[>=3.16 <4]")
@@ -93,13 +102,14 @@ class CycloneDDSCXXConan(ConanFile):
         tc.variables["BUILD_EXAMPLES"] = False
         # variables which effects build
         tc.variables["ENABLE_LEGACY"] = False
-        tc.variables["ENABLE_SHM"] = self.dependencies["cyclonedds"].options.with_shm
+        tc.variables["ENABLE_SHM"] = self.options.with_shm
         tc.variables["ENABLE_TYPE_DISCOVERY"] = self.dependencies["cyclonedds"].options.enable_discovery
         tc.variables["ENABLE_TOPIC_DISCOVERY"] = self.dependencies["cyclonedds"].options.enable_discovery
         tc.variables["ENABLE_COVERAGE"] = False
         tc.generate()
-        cd = CMakeDeps(self)
-        cd.generate()
+        deps = CMakeDeps(self)
+        deps.set_property("iceoryx", "cmake_file_name", "iceoryx_binding_c")
+        deps.generate()
 
     def _patch_sources(self):
         cmakelists = os.path.join(self.source_folder, "CMakeLists.txt")
@@ -154,6 +164,8 @@ class CycloneDDSCXXConan(ConanFile):
         self.cpp_info.components["ddscxx"].set_property("cmake_target_name", "CycloneDDS-CXX::ddscxx")
         self.cpp_info.components["ddscxx"].set_property("pkg_config_name", "CycloneDDS-CXX")
         self.cpp_info.components["ddscxx"].requires = ["cyclonedds::CycloneDDS"]
+        if self.options.with_shm:
+            self.cpp_info.components["ddscxx"].requires.append("iceoryx::iceoryx")
         if self.settings.os in ["Linux", "FreeBSD"]:
             self.cpp_info.components["ddscxx"].system_libs = ["m"]
         self.cpp_info.components["idlcxx"].libs = ["cycloneddsidlcxx"]

--- a/recipes/cyclonedds/all/conanfile.py
+++ b/recipes/cyclonedds/all/conanfile.py
@@ -107,6 +107,7 @@ class CycloneDDSConan(ConanFile):
         tc.variables["BUILD_IDLC_TESTING"] = False
         tc.variables["BUILD_DDSPERF"] = False
         tc.variables["BUILD_IDLC_TESTING"] = False
+        tc.cache_variables["ENABLE_LTO"] = False
         # variables which effects build
         tc.variables["ENABLE_SSL"] = self.options.with_ssl
         tc.variables["ENABLE_SHM"] = self.options.with_shm

